### PR TITLE
Add incompatible flag to disable cfg="host" from Starlark

### PR DESCRIPTION
**Pull request by Yannic**
Host transitions are problematic where the host and exec environment are
different (e.g., with Remote Execution). This change adds an incompatible
flag that fails analysis if Starlark rules use `cfg = "host"`.

Migrating is pretty straigt forward (replace `host` with `exec`) and
should be automatable by buildifier (although I haven't tried to do so).

### CODE REVIEW

**Comment by Yannic** [comment]
@katre and/or @gregestren PTAL

**Comment by katre** [comment]
@comius is actually working on something similar, so I'm assigning this to him so the two of you can discuss.

**Comment by comius** [comment]
cc @susinmotion

I would attempt the more aggressive path first, where we just change all "host" transitions to "exec". This would reduce the churn substantially.

We're already fixing some issues that happen with that plan (within Bazel).

If this plan fails, this PR becomes viable alternative.

**Comment by Yannic** [comment]
In my experiments, it seemed like tools build in exec config do get `--host_*` flags, so that plan doesn't look obviously broken (at least for local exec, didn't try remote exec yet but I should probably do that). SGTM

I think we should probably still make `cfg="host"` an error eventually even after `cfg="host" == cfg="exec"` (similar to what was done to `cfg="data"` which first became a no-op and was the removed), but we can handle that at a later point.

Should I send the documentation change in a separate PR? That seems useful regardless of whether we add the flag or not.

**Comment by gregestren** [comment]
@susinmotion 's been focusing more on this recently. @susinmotion - do you have recommendations for @Yannic 's latest comment?

**Comment by susinmotion** [comment]
Cool, thanks for thinking about this. A separate PR for documentation sounds great--we're really close to making this change.

I'm torn about making cfg="host" an error after the migration. On the one hand, it will make people's rules less confusing in the long term, since they'll say "exec" when they mean "exec". On the other hand, it causes breakages with things that aren't actually a problem, and that's a bummer for users. Can we do some kind of migration script here instead?

**Comment by Yannic** [comment]
@susinmotion thanks for the update!

> Can we do some kind of migration script here instead?

Yes! Added a `buildifier` warning/fix for this: https://github.com/bazelbuild/buildtools/pull/1053

**Comment by comius** [comment]
Hey, can you resolve the conflicts? Host transition has been removed now, so a flag to completely disable it is useful.

**Review by comius — CHANGES_REQUESTED** [review]
Please resolve the conflicts.

**Comment by comius** [comment]
> Please resolve the conflicts.

Friendly ping.

**Comment by Yannic** [comment]
@comius Updated, PTAL

**Review by comius — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/packages/semantics/BuildLanguageOptions.java [hunk 1]
```diff
@@ -532,6 +532,17 @@
...
 
+  @Option(
+      name = "incompatible_disable_starlark_host_transitions",
+      defaultValue = "false",
+      documentationCategory = OptionDocumentationCategory.STARLARK_SEMANTICS,
+      metadataTags = {OptionMetadataTag.INCOMPATIBLE_CHANGE},
+      effectTags = {OptionEffectTag.LOADING_AND_ANALYSIS},
+      help =
+          "If set to true, rule attributes cannot set 'cfg = \"host\"'. Rules should set "
+              + "'cfg = \"host\"' instead.")
```
Comment (comius) [inline src/main/java/com/google/devtools/build/lib/packages/semantics/BuildLanguageOptions.java:543]: Second host should be exec.

Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/packages/semantics/BuildLanguageOptions.java:543]:
```suggestion
              + "'cfg = \"exec\"' instead.")
```

**Comment by brentleyjones** [comment]
@bazel-io flag

**Pull request closed without merging**